### ISSUE TITLE
Remove/avoid side-effect-based multi-instance detection

### ISSUE THREAD

**Issue by mattlubner** [styled-components/styled-components#2987]
## Summary

The multi-instance warning is (misleadingly) logged to the console after the require cache is cleared and `styled-components` is re-imported. There doesn't seem to be a clean / officially-sanctioned way to reset this warning after the require cache is cleared.

The warning in question:
```
It looks like there are several instances of 'styled-components' initialized in this
application. This may cause dynamic styles not rendering properly, errors happening
during rehydration process, missing theme prop, and makes your application bigger
without a good reason.

See https://s-c.sh/2BAXzed for more info.
```

## Reproduction

Unfortunately, `codesandbox.io` doesn't seem to support calling `jest.resetModules()` _or_ `delete require.cache['styled-components']`, and additionally, I was greeted with a wall of error messages when trying to yarn-install a fork of `styled-components`. As a result, I've listed the reproduction steps below:

## Steps to reproduce

_Do the following in a `jest` test:_
1. Set `process.env.NODE_ENV` to any value other than `'production'`
2. Import `styled-components`
3. Reset the environment's require cache: `jest.resetModules()`
4. Re-import `styled-components`
5. Observe the multi-instance warning in the console.

## Expected Behavior

The multi-instance warning shouldn't be logged to the console, since the previously-loaded instance was purged.

## Actual Behavior

The multi-instance warning is still logged, because importing `styled-components` has undocumented side-effects on the javascript environment. Namely, `window['__styled-components-init__']` is used to determine if multiple instances have been loaded. This side-effect survives puring of the require cache, which unfortunately results in a misleading error message being logged to the console.

## Possible Solutions

The easiest solution would be to restore the previous conditional logic guarding this console warning. It appears to have been decimated in this commit: https://github.com/styled-components/styled-components/commit/19dd37a0873ad799f6e48315124a880c4ecd36fc (👋 @probablyup)

Ideally, it would be best to avoid using a side-effect to detect the multi-instance case (perhaps by directly testing `require.cache['styled-components]`), but that wouldn't work in a CDN build (…I think? this is an assumption…). So this case might not cover all scenarios.

Another solution (which I'm totally okay with if y'all are) would be to add the work-around to the [FAQ entry on multiple instances](https://styled-components.com/docs/faqs#why-am-i-getting-a-warning-about-several-instances-of-module-on-the-page) of `styled-components`:

```js
require('./moduleThatImportsStyledComponents');
jest.resetModules();
delete window['__styled-components-init__']; // silences the multi-instance warning
require('./moduleThatImportsStyledComponents');
```

If a simple revert isn't preferred, I'd be happy to make the desired changes, just let me know which approach to take.

---

<details>
  <summary>Environment Details</summary>
<h2>System:</h2>
<ul>
<li>OS: macOS 10.15.2</li>
<li>CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz</li>
<li>Memory: 3.83 GB / 32.00 GB</li>
<li>Shell: 3.2.57 - /bin/bash</li>
</ul>
<h2>Binaries:</h2>
<ul>
<li>Node: 12.14.1 - /usr/local/bin/node</li>
<li>Yarn: 1.21.1 - /usr/local/bin/yarn</li>
<li>npm: 6.13.4 - /usr/local/bin/npm</li>
<li>Watchman: 4.9.0 - /usr/local/bin/watchman</li>
</ul>
<h2>npmPackages:</h2>
<ul>
<li>babel-plugin-styled-components: ^1.10.6 =&gt; 1.10.6</li>
<li>styled-components: ^5.0.0 =&gt; 5.0.0</li>
<li>jest: ^23.6.0 =&gt; 23.6.0</li>
</ul>
</details>

**Comment by mattlubner** [comment]
Whoops, I forgot to paste the console warning in question!

```
It looks like there are several instances of 'styled-components' initialized in this
application. This may cause dynamic styles not rendering properly, errors happening
during rehydration process, missing theme prop, and makes your application bigger
without a good reason.

See https://s-c.sh/2BAXzed for more info.
```

**Comment by mattlubner** [comment]
Yikes… sorry for all the noise! I also meant to include my work-around:

```js
jest.resetModules();
delete window['__styled-components-init__'];
```

Adding the above work-around to the FAQ would be fine, too (assuming it's alright to treat this as an official work-around).

**Comment by probablyup** [comment]
Restoring the `process.env.NODE_ENV !== 'test'` part should be sufficient to get things working in Jest if you want to make a PR against master

**Comment by mattlubner** [comment]
I got sidetracked by a version incompatibility bug (#2989). I'll submit a PR as soon as I have a resolution on that issue.
